src/components/contractSrc.js

```diff
@@ -0,0 +1,24 @@
+export const contractSource = `contract CryptoHamster =
+   datatype event = NewHamster(indexed int, string, hash)
+
+   record state = { hamsters : map(string, hash), next_id : int }
+
+   stateful entrypoint init() = { hamsters = {}, next_id = 0 }
+
+   entrypoint nameExists(name: string) : bool =
+      Map.member(name, state.hamsters)
+
+   stateful entrypoint createHamster(hamsterName: string) =
+      require(!nameExists(hamsterName), "Name is already taken")
+      createHamsterByNameDNA(hamsterName, generateDNA(hamsterName))
+   
+   entrypoint getHamsterDNA(hamsterName: string) : hash =
+      require(nameExists(hamsterName), "Hamster does not exist!")
+      state.hamsters[hamsterName]
+
+   stateful function createHamsterByNameDNA(name: string, dna: hash) =
+      Chain.event(NewHamster(state.next_id, name, dna))
```

Comment:
can you move 'event' at last position? if there is a case where you cannot update/put state with a new hamster, 'event' would be triggered but hamster would not be added. in our case, it's not possible but ...

Reply:
i agree